File: test/language/templates/cpp98/InstantiateTemplateArgumentTest.cpp
#include "gtest/gtest.h"

#include "templates/instantiate_template_argument/InstantiateTemplateArgument.h"

namespace templates
{
namespace instantiate_template_argument
{

TEST(InstantiateTemplateArgumentTest, readWrite)
{
    InstantiateTemplateArgument instantiateTemplateArgument;
    Other_Str& other = instantiateTemplateArgument.getOther();
    Str& str = other.getValue();
    str.setValue("test");

    zserio::BitStreamWriter writer;
    instantiateTemplateArgument.write(writer);
    size_t bufferSize = 0;
    const uint8_t* buffer = writer.getWriteBuffer(bufferSize);

    zserio::BitStreamReader reader(buffer, bufferSize);
    InstantiateTemplateArgument readInstantiateTemplateArgument(reader);

    ASSERT_TRUE(instantiateTemplateArgument == readInstantiateTemplateArgument);
}

} // namespace instantiate_template_argument
} // namespace templates

File: test/language/templates/java/templates/InstantiateTemplateArgumentTest.java
package templates;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

import templates.instantiate_template_argument.InstantiateTemplateArgument;
import templates.instantiate_template_argument.Other_Str;
import templates.instantiate_template_argument.Str;

public class InstantiateTemplateArgumentTest
{
    @Test
    public void readWrite() throws IOException
    {
        final InstantiateTemplateArgument instantiateTemplateArgument =
                new InstantiateTemplateArgument(new Other_Str(new Str("test")));

        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        instantiateTemplateArgument.write(writer);

        final BitStreamReader reader = new ByteArrayBitStreamReader(
                writer.toByteArray(), writer.getBitPosition());
        InstantiateTemplateArgument readInstantiateTemplateArgument = new InstantiateTemplateArgument(reader);
        assertTrue(instantiateTemplateArgument.equals(readInstantiateTemplateArgument));
    }
}
